envoy/stats/store.h

```diff
@@ -27,10 +27,53 @@ class Sink;
 class SinkPredicates;
 
 /**
- * A store for all known counters, gauges, and timers.
+ * Store keeps track of all Scopes created in it, and the Scopes manage
+ * individual stats. Each stat is defined in a scope. There is a single root
+ * scope created in the Store, and more sub-scopes can be created.
+ *
+ * Stores enable iteration over all stats in its owned Scopes.
+ *
+ * There is typically one Store instance in a test or binary, though Isolated
+ * Stores can be created in some scenarios. Stores are typically allocated
+ * as part of other objects or via std::unique_ptr.
+ *
+ * In contast, Scopes are managed via shared_ptr, unique nickname
+ * ScopeSharedPtr, and should not be directly instantiated or allocated via
+ * std::unique_ptr.
+ *
+ * A reference to the root-scope held by the Store until it shuts down. Holding
+ * onto a reference to the root-scope's shared_ptr that outlives the Store is
+ * not allowed.
  */
-class Store : public Scope {
+class Store {
 public:
+  virtual ~Store() = default;
+
+  /**
+   * @return the root scope (creating it if necessary)
+   */
+  virtual ScopeSharedPtr rootScope() PURE;
+
+  /**
+   * @return the root scope (creating it if necessary)
+   */
+  virtual ConstScopeSharedPtr constRootScope() const PURE;
```

Comment:
not sure if this "constXXXX" is a pattern in Envoy. 
I think it's usually just 
const xxx()const

Reply:
I have taken to using it because sometimes the overload is ambiguous for my purposes.